### ISSUE TITLE
python bcc have no destructor

### ISSUE THREAD

**Issue by zaafar** [iovisor/bcc#359]
Hello,

I am trying to add bpf_obj_pin support and want to remove the pinned bpf table when python BPF class is destroyed. I have written the bpf_obj_pin deletion code in BPFModule destructor. However, currently it isn't called at python BPF class destruction.

cc/bpf_module.h have a destructor (BPFModule::~BPFModule()), which is only called in case the BPF code fails to compile. Shouldn't python BPF class need to have a destructor (__del__ or __exit__) which should call the cc/bpf_module.h destructor at exit?

diff --git a/src/python/bcc/__init__.py b/src/python/bcc/__init__.py
index 8c1ef59..8e45154 100644
--- a/src/python/bcc/__init__.py
+++ b/src/python/bcc/__init__.py
@@ -1064,3 +1064,6 @@ class BPF(object):
         except KeyboardInterrupt:
             exit()

+    def __del__(self):
+        if(self.module):
+            return lib.bpf_module_destroy(self.module)
 

or are there some cases where python BPF code exit but we don't want to destroy BPFModule?

**Comment by drzaeus77** [comment]
Can you tell me what your use case for bpf_obj_pin is? Why should it necessarily be tied to the BPF class lifetime? The objects that are pinned are done that way (to the filesystem) so as to provide persistence between processes or for the same process across restarts. Going with an automated destruction approach defeats some of that.

One possible use for pinned objects is to build a reliable service on top of. To do that, there is a bit more orchestration required, possibly from something like upstart or systemd to coordinate/cleanup names, paths, pids, and so on. I don't see the users of the python API bindings getting an immediate use out of that, as all of the functionality so far is being implemented in a single-use script. I have heard of some folks that may be building an agent using the C++ APIs, and such an agent may have such use for pinned objects, but I haven't heard of any requirements from them yet so not sure how to best guide the development of that feature.

On another front, the hover daemon in the iomodules repository may have a similar requirement for persistent objects, but I am writing that in go from scratch and so at best I would get some use out of the C API only.

**Comment by zaafar** [comment]
Yes you’re right that it doesn’t make any sense to unpin BPF objects at BPF class destruction as long as you’re concerned with persistence or reliability. However, our use case is a bit different. We want to share BPF maps/programs across Docker containers by mounting a host directory in containers as data volume. Assuming each container persists its BPF object into this volume, each container BPF module could now tail call into another container’s BPF module bypassing host network stack.

Since doing all this with C API is quite cumbersome, we wanted to leverage bcc’s python binding to accomplish this by writing a patch to support bpf object persistence. Let me know if there’s a better way to get all this done or if this all makes any sense at all.
